### PR TITLE
New can-run/can-save restrictions for metrics

### LINKED ISSUE #47149: [BE] can-run/can-save metric constraints
New constraints:
- 0-1 breakouts
- Breakouts can be on date or datetime columns only

Note - the FE uses `is-date-or-datetime?` helper for type checks. The BE needs to be consistent and do the same.

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/47147
Closes https://github.com/metabase/metabase/issues/47149
Corresponding FE PR https://github.com/metabase/metabase/pull/47368

New metric restrictions:
- 0-1 breakouts
- breakouts only on Date or DateTime columns

### CODE REVIEW

File: test/metabase/lib/query_test.cljc [hunk 1]
```diff
@@ -207,13 +207,23 @@
...
                           (lib/aggregate (lib/sum (meta/field-metadata :venues :id))))
-      true  :metric   (-> lib.tu/venues-query
+      true  :metric   (-> (lib/query meta/metadata-provider (meta/table-metadata :people))
                           (lib/aggregate (lib/count))
-                          (lib/breakout (first (lib/breakoutable-columns lib.tu/venues-query))))
-      false  :metric   (-> lib.tu/venues-query
-                           (lib/aggregate (lib/count))
-                           (lib/append-stage)
-                           (lib/aggregate (lib/count))))))
+                          (lib/breakout (meta/field-metadata :people :created-at)))
```
Comment (ranquild) [inline test/metabase/lib/query_test.cljc:212]: `type/DateTime`

File: test/metabase/lib/query_test.cljc [hunk 2]
```diff
@@ -207,13 +207,23 @@
...
+      true  :metric   (-> (lib/query meta/metadata-provider (meta/table-metadata :people))
+                          (lib/aggregate (lib/count))
+                          (lib/breakout (meta/field-metadata :people :birth-date)))
```
Comment (ranquild) [inline test/metabase/lib/query_test.cljc:215]: `type/Date`

File: src/metabase/lib/query.cljc [hunk 3]
```diff
@@ -63,12 +63,6 @@
   [_query _card-type]
   true)
 
-(defmethod can-run-method :mbql.stage/mbql
```
Comment (ranquild) [inline src/metabase/lib/query.cljc:66]: Circular dependencies

File: src/metabase/lib/stage.cljc [hunk 4]
```diff
@@ -414,3 +416,15 @@
...
+(defmethod lib.query/can-run-method :mbql.stage/mbql
+  [query card-type]
+  (or (not= card-type :metric)
+      (let [aggregations (lib.aggregation/aggregations query 0)
+            breakouts    (lib.breakout/breakouts query 0)]
+        (and (= (lib.query/stage-count query) 1)
+             (= (count aggregations) 1)
+             (or (empty? breakouts)
+                 (and (= (count breakouts) 1)
+                      (-> (lib.breakout/breakout-column query (first breakouts))
```
Comment (ranquild) [inline src/metabase/lib/stage.cljc:429]: Note - `lib.breakout/breakout-column` doesn't change the `:effective-type` of the column when there is an extraction `:temporal-unit` on a breakout but `lib.metadata.calculation/metadata` does. We already depend on this, so I used the same function here and added tests for these cases.

**Review by bshepherdson — APPROVED** [review]

**Review by metamben — APPROVED** [review]
LGTM

**Pull request merged**